### PR TITLE
chore: expose `globalThis.crypto` when not available

### PR DESCRIPTION
**Pull request by balazsorban44**
### What?

Exposing `globalThis.crypto`, based on [Node.js' WebCrypto API](https://nodejs.org/api/globals.html#crypto_1)

### Why?

Similar to `fetch`, `crypto` is a popular API that is currently not available on `globalThis` in all active Node.js versions yet.

This can help library authors to create runtime-agnostic packages.

### How?

Node.js already has the WebCrypto API that can be imported, we just expose it on `globalThis` in Node.js versions where this is not available.

Closes NEXT-1063

[Slack thread](https://vercel.slack.com/archives/C03KAR5DCKC/p1681821510191059)

### CODE REVIEW

File: packages/next/src/server/node-polyfill-crypto.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+if (!(global as any).crypto) {
+  function getCryptoImpl() {
+    return require('node:crypto').webcrypto
+  }
+  Object.defineProperty(global, 'crypto', {
+    get() {
+      return getCryptoImpl()
+    },
+  })
```
Comment (balazsorban44) [inline packages/next/src/server/node-polyfill-crypto.ts:11]: This is done similarly to how we polyfill `fetch`:  https://github.com/vercel/next.js/blob/0918ebd69a23622d78e0f25b2993beea47f5a55f/packages/next/src/server/node-polyfill-fetch.ts#L51-L68

I synced on it with @Ethan-Arrowood but I cannot recall why we needed to use `Object.defineProperty`, but I think it was compatibility/perf related.

File: test/e2e/app-dir/crypto-globally-available/app/handler/route.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export function GET() {
+  return new Response(
+    typeof globalThis.crypto === 'object'
+      ? 'crypto is available'
+      : 'crypto is not available'
+  )
+}
+
+export const runtime = 'nodejs'
```
Comment (balazsorban44) [inline test/e2e/app-dir/crypto-globally-available/app/handler/route.ts:9]: Set explicitly so the test will work indefinitely in the future. Note: `"edge"` already exposes `globalThis.crypto` for us.

File: test/e2e/app-dir/crypto-globally-available/app/page.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  return (
+    <p>
+      {typeof globalThis.crypto === 'object'
+        ? 'crypto is available'
+        : 'crypto is not available'}
+    </p>
+  )
+}
+
+export const runtime = 'nodejs'
```
Comment (balazsorban44) [inline test/e2e/app-dir/crypto-globally-available/app/page.tsx:11]: See: https://github.com/vercel/next.js/pull/48941/files#r1180097586

**Comment by ijjk** [comment]
## Failing test suites 
Commit: f70efc00504fa2da8f27239bd6defea2c8bb26bd

`pnpm testheadless test/e2e/app-dir/crypto-globally-available/crypto-globally-available.test.ts` 
- Web Crypto API is available globally > should be available in Server Components
- Web Crypto API is available globally > should be available in Route Handlers

<details>
<summary>Expand output</summary>

  ● Web Crypto API is available globally › should be available in Server Components

    expect(received).toBe(expected) // Object.is equality

    Expected: "crypto is available"
    Received: "crypto is not available"

      10 |     it('should be available in Server Components', async () => {
      11 |       const browser = await next.browser('/')
    > 12 |       expect(await browser.elementByCss('p').text()).toBe('crypto is available')
         |                                                      ^
      13 |     })
      14 |
      15 |     // In case you need to test the response object

      at Object.<anonymous> (e2e/app-dir/crypto-globally-available/crypto-globally-available.test.ts:12:54)

  ● Web Crypto API is available globally › should be available in Route Handlers

    expect(received).toContain(expected) // indexOf

    Expected substring: "crypto is available"
    Received string:    "crypto is not available"

      17 |       const res = await next.fetch('/handler')
      18 |       const html = await res.text()
    > 19 |       expect(html).toContain('crypto is available')
         |                    ^
      20 |     })
      21 |   }
      22 | )

      at Object.<anonymous> (e2e/app-dir/crypto-globally-available/crypto-globally-available.test.ts:19:20)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: package.json [hunk 4]
```diff
@@ -249,7 +249,7 @@
     "@types/react-dom": "18.0.11"
   },
   "engines": {
-    "node": ">=16"
+    "node": ">=16.8.0"
```
Comment (balazsorban44) [inline package.json:252]: I realized that `webcrypto` was actually added back in `15.0.0` https://nodejs.org/dist/latest-v18.x/docs/api/crypto.html#cryptowebcrypto

But we might still want to bump this version since `undici` (what we use in App Router) requires `16.8.0`.

Let me know if this should be a different PR.

File: packages/next/package.json [hunk 5]
```diff
@@ -315,6 +315,6 @@
     "caniuse-lite": "1.0.30001406"
   },
   "engines": {
-    "node": ">=16"
+    "node": ">=16.8.0"
```
Comment (balazsorban44) [inline packages/next/package.json:318]: See: https://github.com/vercel/next.js/pull/48941/files#r1180126728

File: packages/next/src/server/node-polyfill-crypto.ts (hunk 1 above)
Comment (timneutkens) [reply packages/next/src/server/node-polyfill-crypto.ts:11]: Yeah it's performance related, loading it before it's used adds additional time during cold boot.

**Review by timneutkens — APPROVED** [review]
Looks good to me, thanks Balazs 💯

**Comment by t3dotgg** [comment]
Wait this unbreaks a ton of stuff for me thank you so much
